File: Python Programming/tic-tac-toe.py
import random

the_board = [[".",".","."],[".",".","."],[".",".","."]]

def display_round_number(round_number):
    
    print("\n")
    print("##############")
    print(f"## Round #{round_number} ##")
    print("##############")
    print("\n")
    
def display_board(the_board): 
    
    print("Current Board: ")
    print("~~~~~~~~~~~~")
    print("   " + " 0 " + " 1 " + " 2 ")
    print(" 0 " + " " + the_board[0][0] + " " + " " + the_board[0][1] + " " + " "+ the_board[0][2] + " ")
    print(" 1 " + " " + the_board[1][0] + " " + " " + the_board[1][1] + " " + " "+ the_board[1][2] + " ")
    print(" 2 " + " " + the_board[2][0] + " " + " " + the_board[2][1] + " " + " "+ the_board[2][2] + " ")
    print("~~~~~~~~~~~~")

def player_input():
    
    marker = input("X or O? ").upper()

    while not (marker == 'X' or marker == 'O'):
        print("Invalid response. Choose X or O.")
        marker = input("X or O? ").upper()
    
    if marker == 'X':
        return ('X', 'O')
    else:
        return ('O', 'X')

def place_marker(board, marker, position):
    
    board[position[0]][position[1]] = marker

def win_check(board, mark):
    
    is_win = (the_board[0][0] == the_board[0][1] == the_board[0][2] == mark) or (the_board[1][0] == the_board[1][1] == the_board[1][2] == mark) or (the_board[2][0] == the_board[2][1] == the_board[2][2] == mark) or (the_board[0][0] == the_board[1][0] == the_board[2][0] == mark) or (the_board[0][1] == the_board[1][1] == the_board[2][1] == mark) or (the_board[0][2] == the_board[1][2] == the_board[2][2] == mark) or (the_board[0][0] == the_board[1][1] == the_board[2][2] == mark) or (the_board[0][2] == the_board[1][1] == the_board[2][0] == mark)
    
    return is_win

def space_check(board, position):
    
    return board[position[0]][position[1]] == "."
    

def full_board_check(board):
    
    for row_num in range(0, 2):
        for column_num in range(0, 2):
            if board[row_num][column_num] == '.':
                return False
    
    return True

def player_position(marker, board):

    repeat_again = True
    
    while repeat_again == True:
        row_number = input("What row? ")
        column_number = input("What column? ")
        
        try: 
            row_number = int(row_number)
            column_number = int(column_number)
        except ValueError:
            print('Invalid input. Choose 0, 1, or 2 for row and column.')
        else:
            position = [row_number, column_number]
            if (row_number not in range(3)) or (column_number not in range(3)) or (space_check(board, position) == False):
                print('Invalid location. Choose 0, 1, or 2 for row and column.')
            else:
                check_pos = input(f"Move '{marker}' at row {row_number} and column {column_number}. y/n? ").lower()
                if check_pos == 'y':
                    print("Move placed!")
                    print("\n")
                    repeat_again = False
                elif check_pos == 'n':
                    print("Move not placed.")
                else:
                    print("Invalid input. Try again.")
      
    return position

def computer_position(board):
    
    repeat_again = True
    
    while repeat_again == True:
        row_number = random.randint(0,2)
        column_number = random.randint(0,2)
        position = [row_number, column_number]
        if space_check(board, position) == True: 
            repeat_again = False
            print('Computer move placed.')
            print('\n')
    
    return position

def game():

    game_over = False
    
    round_number = 1
    marker_choice = player_input()
    player = marker_choice[0]
    computer = marker_choice[1]
    rotation = 2
    
    if player == 'X' and computer == 'O':
        player_turn = True
    elif player == 'O' and computer == 'X':
        player_turn = False
    
    
    while not game_over: 
        
        if rotation % 2 == 0: 
            display_round_number(round_number)
            display_board(the_board)
            round_number += 1
     
        if player_turn == True:
            print("\n")
            print(f"Player {player} turn.")
            player_pos = player_position(player, the_board)
            place_marker(the_board, player, player_pos)
            display_board(the_board)
            player_win = win_check(the_board, player)
            full_board = full_board_check(the_board)
            if player_win == True: 
                print("You win!")
                game_over = True
            else:
                if full_board == True:
                    print("It's a draw.")
                    game_over = True
                else:
                    player_turn = False
            rotation += 1
        else:
            print("\n")
            print(f"Player {computer} turn.")
            computer_pos = computer_position(the_board)
            place_marker(the_board, computer, computer_pos)
            display_board(the_board)
            computer_win = win_check(the_board, computer)
            full_board = full_board_check(the_board)
            if computer_win == True: 
                print("You lose!")
                game_over = True
            else:
                if full_board == True:
                    print("It's a draw.")
                    game_over = True
                else:
                    player_turn = True
            rotation += 1

game()
               
    

File: R Programming/tic-tac-toe.R
### GAME FUNCTIONS ###

# display_round_number() displays the round number
display_round_number <- function(round_num){
  
  cat("###############", "\n")
  cat("### Round", round_num, "###", "\n")
  cat("###############\n")
  
}

# display_board() displays the board
display_board <- function(board){
  
  cat("Current Board:", "\n\n", "~~~~~~~~~~~~~~~~~~~", "\n\n")
  print(board)
  cat("\n\n", "~~~~~~~~~~~~~~~~~~~", "\n\n")
  
}

# player_input() prompts the user to choose between X or O
player_input <- function(con) {
  
  cat("X or O? ")
  marker <- toupper(readLines(con=con, n=1))
  
  while (!marker %in% c("X", "O")) {
    cat("Invalid response. Choose X or O.")
    marker <- toupper(readLines(con=con, n=1))
  }
  
  if (marker == "X"){
    return(c("X", "O"))
  } else {
    return(c("O", "X"))
  }
}

# place_symbol() updates the board 
place_symbol <- function(board, symbol, pos){
  board[pos[1], pos[2]] <- symbol
  return(board)
}

# win_check() checks if winning combination exists
win_check <- function(board, symbol){
  
  for(i in 1:length(winning_combinations)){
    if(sum(which(board == symbol) %in% winning_combinations[[i]]) >= 3){
      return(TRUE)
    }
  }
  return(FALSE)
}

# space_check() checks if space in board is empty
space_check <- function(board, pos){
  has_space = is.na(board[pos[1], pos[2]])
  return(has_space)
}

# full_board_check() checks if board is full
full_board_check <- function(board){
  
  not_full = any(is.na(board))
  return(not_full)
}

# player_position() prompts user for location to put marker on board
player_position <- function(board, symbol, con){
  
  while (TRUE){
    
    cat("Please enter row: ")
    row <- as.integer(readLines(con=con, n=1))
    cat("Please enter column: ")
    col <- as.integer(readLines(con=con, n=1))
    pos <- c(row, col)
    
    
    if ((!row %in% seq(1:3)) | (!col %in% seq(1:3))){
      cat("Invalid input. Please choose 1, 2, or 3 for row and column.\n")
      next()
    } else {
      if(space_check(board, pos) == FALSE){
        cat("Invalid location. Please choose 1, 2, or 3 for row and column.\n")
        next()
      } else{
        cat(paste0("Move ", symbol, " at row ", row, " and columns ", col, ". y/n? "))
        check_pos <- tolower(readLines(con=con, n=1))
        if(check_pos == 'y'){
          cat("Move placed!", "\n\n")
          break()
        } else if (check_pos == 'n') {
          cat("Move not placed.", "\n\n")
          next()
        } else {
          cat("Invalid input. Try again.", "\n\n")
          next()
        }
      }
    }
  }
  return(c(row, col))
}

# computer_position() generates a random location on board for computer
computer_position <- function(board){
  
  while (TRUE){
    row <- sample.int(3, 1)
    col <- sample.int(3, 1)
    pos <- c(row, col)
    
    if (space_check(board, pos) == TRUE){
      cat("Computer move placed.", "\n\n")
      break()
    }
  }
  return (pos)
}

# game() runs tic tac toe game
game <- function() {
  
  cat("TIC-TAC TOE GAME!\n")
  Sys.sleep(1)
  
  if (interactive()) {
    con <- stdin()
  } else {
    con <- "stdin"
  }
  
  marker_choice <- player_input(con = con)
  player <- marker_choice[1]
  computer <- marker_choice[2]
  round_num <- 1
  rotation <- 2
  
  if (player == "X" && computer == "O"){
    player_turn = TRUE
  } else {
    player_turn = FALSE
  }
  
  while (TRUE){
    
    if (rotation %% 2 == 0){
      display_round_number(round_num)
      round_num <- round_num + 1
      Sys.sleep(1)
    }
    
    if (player_turn == TRUE) {
      cat("Player ", player, " turn.", "\n\n")
      player_pos = player_position(board, player, con)
      board <- place_symbol(board, player, player_pos)
      display_board(board)
      player_win <- win_check(board, player)
      not_draw <- full_board_check(board)
      
      if (player_win == TRUE){
        cat("Congratulations! You Win!\n")
        break()
      } else{
        if (not_draw == FALSE){
          cat("It's a draw.\n")
          break()
        } else{
          player_turn <- FALSE
          rotation <- rotation + 1
          Sys.sleep(1)
        }
      }
    } else {
      cat("Player ", computer, " turn.", "\n\n")
      computer_pos = computer_position(board)
      board <- place_symbol(board, computer, computer_pos)
      display_board(board)
      computer_win = win_check(board, computer)
      not_draw = full_board_check(board)
      
      if (computer_win == TRUE) {
        print("You lose.\n")
        break()
      } else {
        if (not_draw == FALSE) {
          cat("It's a draw.\n")
          break()
        } else {
          player_turn <- TRUE
          rotation <- rotation + 1
          Sys.sleep(1)
        }
      }
    }
    }
  }

#### Variables and Game ####
board <- matrix(nrow=3, ncol=3)
winning_combinations = list(c(1,2,3), c(4,5,6), c(7,8,9), c(1,4,7), c(2,5,8), c(3,6,9), c(1,5,9), c(3,5,7))
game()
